isisd/isis_route.c

```diff
@@ -239,41 +239,51 @@ static void isis_route_info_delete(struct isis_route_info *route_info)
 	XFREE(MTYPE_ISIS_ROUTE_INFO, route_info);
 }
 
-static int isis_route_info_same_attrib(struct isis_route_info *new,
-				       struct isis_route_info *old)
-{
-	if (new->cost != old->cost)
-		return 0;
-	if (new->depth != old->depth)
-		return 0;
-
-	return 1;
-}
-
 static int isis_route_info_same(struct isis_route_info *new,
-				struct isis_route_info *old, uint8_t family)
+				struct isis_route_info *old, char *buf,
+				size_t buf_size)
 {
 	struct listnode *node;
 	struct isis_nexthop *nexthop;
 
-	if (!CHECK_FLAG(old->flag, ISIS_ROUTE_FLAG_ZEBRA_SYNCED))
+	if (new->cost != old->cost) {
+		if (buf)
+			snprintf(buf, buf_size, "cost (old: %u, new: %u)",
+				 old->cost, new->cost);
 		return 0;
+	}
 
-	if (CHECK_FLAG(new->flag, ISIS_ROUTE_FLAG_ZEBRA_RESYNC))
+	if (new->depth != old->depth) {
+		if (buf)
+			snprintf(buf, buf_size, "depth (old: %u, new: %u)",
+				 old->depth, new->depth);
 		return 0;
+	}
 
-	if (!isis_route_info_same_attrib(new, old))
+	if (new->nexthops->count != old->nexthops->count) {
+		if (buf)
+			snprintf(buf, buf_size, "nhops num (old: %u, new: %u)",
+				 old->nexthops->count, new->nexthops->count);
 		return 0;
+	}
 
-	for (ALL_LIST_ELEMENTS_RO(new->nexthops, node, nexthop))
+	for (ALL_LIST_ELEMENTS_RO(new->nexthops, node, nexthop)) {
 		if (!nexthoplookup(old->nexthops, nexthop->family, &nexthop->ip,
-				   nexthop->ifindex))
+				   nexthop->ifindex)) {
+			if (buf)
+				snprintf(buf, buf_size,
+					 "new nhop"); /* TODO: print nhop */
 			return 0;
+		}
+	}
 
-	for (ALL_LIST_ELEMENTS_RO(old->nexthops, node, nexthop))
-		if (!nexthoplookup(new->nexthops, nexthop->family, &nexthop->ip,
-				   nexthop->ifindex))
```

Comment:
Yeah, if the number of nexthops didn't change and there's no new nexthop (previous loop), then there's no need to check if any nexthop was removed :+1: